ruby/ext/google/protobuf_c/defs.c

```diff
@@ -46,29 +48,261 @@ static VALUE rb_str_maybe_null(const char* s) {
+                                 google_protobuf_FileDescriptorProto* file,
+                                 google_protobuf_FieldDescriptorProto* field) {
+  /* Look for TYPE_ENUM fields that have a default. */
+  if (google_protobuf_FieldDescriptorProto_type(field) !=
+          google_protobuf_FieldDescriptorProto_TYPE_ENUM ||
+      !google_protobuf_FieldDescriptorProto_has_default_value(field) ||
+      !google_protobuf_FieldDescriptorProto_has_type_name(field)) {
+    return;
   }
-  return (upb_def*)def;
-}
 
-static upb_msgdef* check_msg_notfrozen(const upb_msgdef* def) {
-  return upb_downcast_msgdef_mutable(check_notfrozen((const upb_def*)def));
-}
+  upb_strview defaultval =
+      google_protobuf_FieldDescriptorProto_default_value(field);
+  upb_strview type_name = google_protobuf_FieldDescriptorProto_type_name(field);
+
+  if (defaultval.size == 0 || !isdigit(defaultval.data[0])) {
+    return;
+  }
+
+  if (type_name.size == 0 || type_name.data[0] != '.') {
+    return;
+  }
+
+  const char *type_name_str = type_name.data + 1;
 
-static upb_fielddef* check_field_notfrozen(const upb_fielddef* def) {
-  return upb_downcast_fielddef_mutable(check_notfrozen((const upb_def*)def));
+  char *end;
+  errno = 0;
+  long val = strtol(defaultval.data, &end, 10);
+
+  if (errno != 0 || *end != 0 || val < INT32_MIN || val > INT32_MAX) {
+    return;
+  }
+
+  /* Now find the corresponding enum definition. */
+  const upb_enumdef *e = upb_symtab_lookupenum(symtab, type_name_str);
+  if (e) {
+    /* Look in previously loaded files. */
+    const char *label = upb_enumdef_iton(e, val);
+    if (!label) {
+      return;
+    }
+    google_protobuf_FieldDescriptorProto_set_default_value(
+        field, upb_strview_makez(label));
+  } else {
+    /* Look in enums defined in this file. */
+    const google_protobuf_EnumDescriptorProto* matching_enum = NULL;
+    size_t i, n;
+    const google_protobuf_EnumDescriptorProto* const* enums =
+        google_protobuf_FileDescriptorProto_enum_type(file, &n);
+    for (i = 0; i < n; i++) {
+      if (upb_strview_eql(google_protobuf_EnumDescriptorProto_name(enums[i]),
+                          upb_strview_makez(type_name_str))) {
+        matching_enum = enums[i];
+        break;
+      }
+    }
+
+    if (!matching_enum) {
+      return;
+    }
+
+    const google_protobuf_EnumValueDescriptorProto* const* values =
+        google_protobuf_EnumDescriptorProto_value(matching_enum, &n);
+    for (i = 0; i < n; i++) {
+      if (google_protobuf_EnumValueDescriptorProto_number(values[i]) == val) {
+        google_protobuf_FieldDescriptorProto_set_default_value(
+            field, google_protobuf_EnumValueDescriptorProto_name(values[i]));
+        return;
+      }
+    }
+
+    /* We failed to find an enum default.  But we'll just leave the enum
+     * untouched and let the normal def-building code catch it. */
+  }
+}
+
+/* Historically we allowed the default to be specified as a number.  In
```

Comment:
What does the default mean here?

Reply:
Enum defaults. I clarified the comment a little.